src/orange/detail/BIHMaker.cc

```diff
@@ -0,0 +1,329 @@
+{
+    CELER_EXPECT(!indices.empty());
+    CELER_EXPECT(!centers.empty());
+    CELER_EXPECT(indices.size() == centers.size());
+
+    VecIndices left;
+    VecIndices right;
+
+    for (auto i : range(indices.size()))
+    {
+        if (centers[i][to_int(p.axis)] < p.location)
+        {
+            left.push_back(indices[i]);
+        }
+        else
+        {
+            right.push_back(indices[i]);
+        }
+    }
+
+    return std::make_pair(left, right);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add leaf volume ids to a given node.
+ */
+void BIHMaker::make_leaf(BIHNode& node, VecIndices const& indices)
+{
+    CELER_EXPECT(!node);
+    CELER_EXPECT(!indices.empty());
+
+    node.vol_ids
+        = make_builder(storage_).insert_back(indices.begin(), indices.end());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the centers of each bounding box.
+ */
+BIHMaker::VecReal3 BIHMaker::centers(VecIndices const& indices)
+{
+    CELER_EXPECT(!indices.empty());
+    VecReal3 centers(indices.size());
+    for (auto i : range(indices.size()))
+    {
+        Real3 center;
+        auto bbox = bboxes_[indices[i].unchecked_get()];
+
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            center[ax] = (bbox.lower()[ax] + bbox.upper()[ax]) / 2;
+        }
+        centers[i] = center;
+    }
+    return centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create sorted and uniquified X, Y, Z values of bbox centers.
+ */
+BIHMaker::AxesCenters BIHMaker::axes_centers(VecReal3 const& centers)
+{
+    CELER_EXPECT(!centers.empty());
+
+    AxesCenters axes_centers{{}, {}, {}};
+
+    for (auto center : centers)
+    {
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            axes_centers[ax].push_back(center[ax]);
+        }
+    }
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+        sort_and_uniquify(axes_centers[ax]);
+    }
+
+    return axes_centers;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Bounding box of a collection of bounding boxes.
+ */
+BoundingBox BIHMaker::meta_bbox(VecIndices const& indices)
+{
+    CELER_EXPECT(!indices.empty());
+
+    auto inf = std::numeric_limits<double>::infinity();
+
+    Real3 lower = {inf, inf, inf};
+    Real3 upper = {-inf, -inf, -inf};
+
+    for (auto id : indices)
+    {
+        auto bbox = bboxes_[id.unchecked_get()];
+        for (auto axis : range(Axis::size_))
+        {
+            auto ax = to_int(axis);
+            lower[ax] = std::min(lower[ax], bbox.lower()[ax]);
+            upper[ax] = std::max(upper[ax], bbox.upper()[ax]);
+        }
+    }
+
+    return {lower, upper};
+}
```

Comment:
This is just a union operation. I think a 
```c++
inline BoundingBox union(BoundingBox const& a, BoundingBox const& b);
```
would be a good candidate for a standalone function in a `BoundingBoxUtils.hh` class.

Reply:
Why that API? It needs to be union over an arbitrary number of bounding boxes, so it should probably take a vector. Also, I am assuming you mean that BoundingBoxUtils.hh should be a collection of free-standing functions, rather than an actual class.

Reply:
Well, a union over an arbitrary number won't be any different than consecutively unioning a bunch of pairs of bboxes (or unioning a single one with a bunch of others). Since we don't expect to be doing this at runtime, performance isn't an issue (and even then I be the compiler could optimize it to the same thing), and the 'union two bboxes' is the most general interface.